### PR TITLE
feat: implement maps in BAML

### LINKED ISSUE #787: BAML language: maps
BAML currently doesn't support map types. Add support for it.

### PR DESCRIPTION
**Pull request by sxlijin**
Add support for maps in the BAML language. We currently only support maps with string keys.

Here's the work item list that we verified (fortunately, most of these were already pre-implemented):

- integ-tests (python, ts, ruby)
- parser for the compiler
    - grammar
    - schema-ast
- parser for codemirror
    - lezer
- vscode: textmate grammar
- diagnostics
    - not sure what this means concretely
- in-Rust implementation, i.e. BamlValue
- value syntax
    - ie how you define a map in a test
- input layer (lang-native object → BamlValue)
- ctx.output_format
- output layer (BamlValue → lang-native object)
- codegen (python, ts, ruby)

We also added:

- a new entry point for `validate_type_allowed` where we apply refinements to the lexed types
- textmate grammar snapshot testing using vscode-tmgrammar-snap

Fixes #787

### CODE REVIEW

File: engine/baml-lib/jsonish/src/tests/test_maps.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
+use super::*;
```
Comment (hellovai) [inline engine/baml-lib/jsonish/src/tests/test_maps.rs:3]: @sxlijin some more tests would be nice

map with new lines, for keys,

partial maps
nested partial maps
maps with ints as keys to ensure they cast to strings
maps with arrays as key to ensure they cast to strings
maps in json markup
nested maps with new lines
union of map vs class (would the map correctly let the class have it?)

File: engine/baml-lib/baml/tests/validation_files/class/map_types.baml [hunk 2]
```diff
@@ -0,0 +1,64 @@
+// Tests for validating map types
```
Comment (hellovai) [inline engine/baml-lib/baml/tests/validation_files/class/map_types.baml:1]: nit add a test for return type in a function as well

File: engine/baml-lib/baml/tests/validation_files/class/secure_types.baml [hunk 3]
```diff
@@ -17,207 +17,45 @@ class ComplexTypes {
   o (((int | string) | bool[]), (float, double) | long_long_identifier_123)
 }
 
-// error: Type `apple_pie` does not exist. Did you mean one of these: `ComplexTypes`, `float`, `bool`, `string`, `int`?
+// error: Error validating: This line is not a valid field or attribute definition. A valid class property looks like: 'myProperty string[] @description("This is a description")'
```
Comment (hellovai) [inline engine/baml-lib/baml/tests/validation_files/class/secure_types.baml]: this error got much worse.

seems we broke the grammar :(

<blocking>

**Review by hellovai — COMMENTED** [review]
almost, just one grammar bug it seems.

also what does the prompt look like for a map type? I don't see how the code changes for that

File: integ-tests/baml_src/test-files/functions/input/named-args/single/named-map-string-to-class.baml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    Return this value back to me: {{myMap}}
+  "#
+}
+
+test TestFnNamedArgsSingleMapStringToClass {
+  functions [TestFnNamedArgsSingleMapStringToClass]
+  args {
+    myMap {
+      "key" "example string"
+    }
```
[the remark this answers is not in the document]
Comment (aaronvg) [reply integ-tests/baml_src/test-files/functions/input/named-args/single/named-map-string-to-class.baml]: wait isnt greptile correct here?


myMap {
   "key" { word "example string" }
 }

File: engine/baml-lib/baml/tests/validation_files/class/secure_types.baml (hunk 3 above)
Comment (sxlijin) [reply engine/baml-lib/baml/tests/validation_files/class/secure_types.baml]: you reviewed an outdated version of the PR - the grammar wasn't broken, the test was still using the `{key: value}` map syntax

**Comment by sxlijin** [comment]
todo: lex maps of arbitrary parity

**Pull request merged**